Modifier keys and mouse events sent to backgrounded Safari window

rr-codes opened the pull request:
#### fe2a6a572c960334e5bd9af22bf151a91e0ad36e
<pre>
Modifier keys and mouse events sent to backgrounded Safari window
<a href="https://bugs.webkit.org/show_bug.cgi?id=274877">https://bugs.webkit.org/show_bug.cgi?id=274877</a>
<a href="https://rdar.apple.com/115430109">rdar://115430109</a>

Reviewed by Simon Fraser.

Currently, the event monitor which dispatches mouse events on modifier key presses gets installed when added
to a window, and removed when removed from a window. However, this results in the monitor still existing for
backgrounded windows.

Fix by creating a new tracking area and ensuring that the monitor is only active while the mouse is within
the tracking area of the window.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::secondaryTrackingAreaOptions):
(WebKit::m_flagsChangedEventMonitorTrackingArea):
(WebKit::WebViewImpl::viewDidMoveToWindow):
(WebKit::WebViewImpl::mouseEntered):
(WebKit::WebViewImpl::mouseExited):
(WebKit::m_primaryTrackingArea): Deleted.

Canonical link: <a href="https://commits.webkit.org/279557@main">https://commits.webkit.org/279557@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6de2f5a3dcf8f411d43e4298e8553777d1b9e17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33145 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6295 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57057 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4502 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56082 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4352 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43547 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2943 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55876 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31367 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46519 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24683 "Found 1 new API test failure: /WPE/TestMultiprocess:/webkit/WebKitWebView/multiprocess-create-ready-close (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28196 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3824 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2657 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4009 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58652 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28944 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4056 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50960 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30140 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46665 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50296 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11708 "Found 1 new test failure: workers/wasm-hashset-many.html (failure)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31074 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7963 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/29920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -1204,6 +1204,11 @@ static NSTrackingAreaOptions trackingAreaOptions()
     return options;
 }
 
+static NSTrackingAreaOptions secondaryTrackingAreaOptions()
```
rename to `flagsChangedEventMonitorTrackingAreaOptions`

aprotyas commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5634,6 +5629,15 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
     if (m_ignoresMouseMoveEvents)
         return;
 
+    if (!m_flagsChangedEventMonitor && event.trackingArea == m_flagsChangedEventMonitorTrackingArea.get()) {
```
Hmm, I would've expected us to do this work in an observer like `windowDidBecomeKey`, and then have the monitor removed in `windowDidResignKey`. Could that help with the test failure?

rr-codes replied:
no, because that would break click through

rr-codes committed:
Modifier keys and mouse events sent to backgrounded Safari window
https://bugs.webkit.org/show_bug.cgi?id=274877
rdar://115430109

Reviewed by Simon Fraser.

Currently, the event monitor which dispatches mouse events on modifier key presses gets installed when added
to a window, and removed when removed from a window. However, this results in the monitor still existing for
backgrounded windows.

Fix by creating a new tracking area and ensuring that the monitor is only active while the mouse is within
the tracking area of the window.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::secondaryTrackingAreaOptions):
(WebKit::m_flagsChangedEventMonitorTrackingArea):
(WebKit::WebViewImpl::viewDidMoveToWindow):
(WebKit::WebViewImpl::mouseEntered):
(WebKit::WebViewImpl::mouseExited):
(WebKit::m_primaryTrackingArea): Deleted.

Canonical link: https://commits.webkit.org/279557@main
